### PR TITLE
optimize seek write-cf

### PR DESCRIPTION
**Pull request by lishihai9017**
the main optimization is using prefix seek instead of total order seek which is already enabled before this pull.
BUT replacing upper bound with prefix bound still can reduce some iterate ops like much del key exist before the bound.
eg. you del a big range of data, and you just want to get the bound value of the deleted range which is not exist.

### CODE REVIEW

File: src/raftstore/store/engine.rs [hunk 1]
```diff
@@ -181,6 +181,10 @@ impl IterOption {
         self
     }
 
+    pub fn prefix_seek_used(&self) -> bool {
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/engine.rs]: prefix_seek != prefix_same_as_start

**Comment by ngaut** [comment]
Please add tests.

File: src/storage/mvcc/reader.rs [hunk 2]
```diff
@@ -149,10 +149,8 @@ impl<'a> MvccReader<'a> {
                 self.write_cursor = Some(iter);
             }
         } else {
-            let upper_bound_key = key.append_ts(0u64);
-            let upper_bound = upper_bound_key.encoded().clone();
             // use prefix bloom filter
-            let iter_opt = IterOption::new(Some(upper_bound), true).use_prefix_seek();
+            let iter_opt = IterOption::new(None, true).use_prefix_seek().enable_prefix_bound();
```
Comment (andelf) [inline src/storage/mvcc/reader.rs]: use default() instead of new().
Actually, `builder style` is aimed to replace `new()` with long parameters.

**Comment by lishihai9017** [comment]
test case `test_write_imp` of txn logic has already cover the function of seek_write_impl. @ngaut

**Comment by lishihai9017** [comment]
the main optimization is using prefix seek instead of total order seek which is already enabled before this pull : (
but replacing upper bound with prefix bound simply the seek logic.

File: src/raftstore/store/engine.rs [hunk 3]
```diff
@@ -163,6 +163,7 @@ enum SeekMode {
 
 pub struct IterOption {
     upper_bound: Option<Vec<u8>>,
+    prefix_bound: bool,
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/engine.rs]: s/prefix_bound/prefix_same_as_start

Comment (zhangjinpeng1987) [inline src/raftstore/store/engine.rs]: Keep consistent with ReadOptions

File: src/storage/mvcc/reader.rs [hunk 4]
```diff
@@ -149,10 +149,8 @@ impl<'a> MvccReader<'a> {
                 self.write_cursor = Some(iter);
             }
         } else {
-            let upper_bound_key = key.append_ts(0u64);
-            let upper_bound = upper_bound_key.encoded().clone();
             // use prefix bloom filter
-            let iter_opt = IterOption::new(Some(upper_bound), true).use_prefix_seek();
+            let iter_opt = IterOption::default().use_prefix_seek().enable_prefix_bound();
```
Comment (zhangjinpeng1987) [inline src/storage/mvcc/reader.rs]: s/enable_prefix_bound/set_prefix_same_as_start

File: src/raftstore/store/engine.rs [hunk 5]
```diff
@@ -193,12 +195,18 @@ impl IterOption {
         self.upper_bound = Some(bound);
         self
     }
+
+    pub fn enable_prefix_same_as_start(mut self) -> IterOption {
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/engine.rs]: s/enable_prefix_same_as_start/set_prefix_same_as_start

File: src/raftstore/store/engine.rs [hunk 6]
```diff
@@ -280,7 +288,11 @@ impl Iterable for DB {
     fn new_iterator(&self, iter_opt: IterOption) -> DBIterator {
         let mut readopts = ReadOptions::new();
         readopts.fill_cache(iter_opt.fill_cache);
-        readopts.set_total_order_seek(iter_opt.total_order_seek_used());
+        if iter_opt.total_order_seek_used() {
```
Comment (siddontang) [inline src/raftstore/store/engine.rs]: should extract there to a function

File: src/raftstore/store/engine.rs [hunk 7]
```diff
@@ -278,23 +300,12 @@ impl Peekable for DB {
 
 impl Iterable for DB {
     fn new_iterator(&self, iter_opt: IterOption) -> DBIterator {
```
Comment (andelf) [inline src/raftstore/store/engine.rs]: 🤣 so why not change IterOption to ReadOption? 
We've made this API a mass

File: src/raftstore/store/engine.rs [hunk 8]
```diff
@@ -193,12 +195,32 @@ impl IterOption {
         self.upper_bound = Some(bound);
         self
     }
+
+    pub fn set_prefix_same_as_start(mut self) -> IterOption {
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/engine.rs]: set_prefix_same_as_start(mut self, v: bool)

File: src/raftstore/store/engine.rs [hunk 9]
```diff
@@ -193,12 +195,32 @@ impl IterOption {
...
+        self.prefix_same_as_start = true;
+        self
+    }
+
+    pub fn to_read_opts(&self) -> ReadOptions {
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/engine.rs]: build_read_opts

**Comment by zhangjinpeng1987** [comment]
LGTM

**Comment by andelf** [comment]
LGTM.

**Pull request merged**